File: app/src/main/java/com/nicholasdoglio/eyebleach/ui/about/AboutAdapter.java
/*
    Aww Gallery
    Copyright (C) 2017  Nicholas Doglio

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.nicholasdoglio.eyebleach.ui.about;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.nicholasdoglio.eyebleach.R;
import com.nicholasdoglio.eyebleach.data.model.about.AboutInfo;
import com.nicholasdoglio.eyebleach.util.Intents;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author Nicholas Doglio
 */
public class AboutAdapter extends RecyclerView.Adapter<AboutAdapter.ViewHolder> {
    private Context aboutContext;

    private List<AboutInfo> aboutList = new ArrayList<>();

    AboutAdapter(Context aboutContext) {
        this.aboutContext = aboutContext;
        populateList(aboutContext);
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.activity_about_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.aboutContentImageView.setImageResource(aboutList.get(position).getImageID());
        holder.aboutContentDescriptionImageView.setText(aboutList.get(position).getContentName());
    }

    @Override
    public int getItemCount() {
        return aboutList.size();
    }

    private void populateList(Context context) {
        aboutList.add(new AboutInfo(R.drawable.ic_info, Intents.INSTANCE.provideVersion(context), "https://github.com/WhosNickDoglio/Aww-Gallery/releases"));
        aboutList.add(new AboutInfo(R.drawable.ic_feedback, "Feedback", ""));
        aboutList.add(new AboutInfo(R.drawable.ic_info, "Libraries", "https://htmlpreview.github.io/?https://github.com/WhosNickDoglio/Aww-Gallery/blob/master/libaries.html"));
        aboutList.add(new AboutInfo(R.drawable.ic_github, "Source code", "https://github.com/WhosNickDoglio/Aww-Gallery"));
        aboutList.add(new AboutInfo(R.drawable.dev_photo, "Developed by Nicholas Doglio", "https://whosnickdoglio.github.io/"));
        aboutList.add(new AboutInfo(R.drawable.designer_photo, "Icon & Graphics by Guzman Gonzalez", "http://guzzgonzalez.com/"));
    }

    class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        @BindView(R.id.content_photo)
        ImageView aboutContentImageView;
        @BindView(R.id.content_description)
        TextView aboutContentDescriptionImageView;

        ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            if (aboutList.get(getAdapterPosition()).getContentName().equalsIgnoreCase("Feedback")) {
                Intents.INSTANCE.composeEmail(aboutContext);
            } else {
                Intents.INSTANCE.openWebPage(aboutContext, aboutList.get(getAdapterPosition()).getContentLink());

            }
        }
    }
}

File: app/src/main/java/com/nicholasdoglio/eyebleach/ui/about/AboutAdapter.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Nicholas Doglio
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.nicholasdoglio.eyebleach.ui.about

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.navigation.NavController
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.nicholasdoglio.eyebleach.R
import com.nicholasdoglio.eyebleach.data.about.AboutInfo
import com.nicholasdoglio.eyebleach.data.about.OpenAction
import com.nicholasdoglio.eyebleach.ui.base.AwwGalleryHolder
import com.nicholasdoglio.eyebleach.ui.util.openWebPage
import kotlinx.android.synthetic.main.item_about.*

class AboutAdapter(private val navController: NavController) :
    ListAdapter<AboutInfo, AboutAdapter.AboutViewHolder>(diff) {
    companion object {
        private val diff = object : DiffUtil.ItemCallback<AboutInfo>() {
            override fun areItemsTheSame(oldItem: AboutInfo, newItem: AboutInfo): Boolean =
                oldItem.name == newItem.name

            override fun areContentsTheSame(oldItem: AboutInfo, newItem: AboutInfo): Boolean =
                oldItem == newItem
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AboutViewHolder =
        AboutViewHolder(
            LayoutInflater.from(parent.context).inflate(R.layout.item_about, parent, false)
        )

    override fun onBindViewHolder(holder: AboutViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class AboutViewHolder(override val containerView: View) : AwwGalleryHolder<AboutInfo>(containerView) {

        override fun bind(model: AboutInfo) {
            content.text = content.resources.getString(model.name)
            containerView.setOnClickListener {
                when (model.action) {
                    is OpenAction.OpenLibs -> navController.navigate(R.id.open_libs)
                    is OpenAction.OpenUrl -> it.context.openWebPage(it.context.getString(model.action.url))
                }
            }
        }
    }
}
